seastar/src/submit_to.rs

```diff
@@ -0,0 +1,110 @@
+use crate as seastar;
+use ffi::*;
+use std::future::Future;
+
+use crate::cxx_async_local_future::IntoCxxAsyncLocalFuture;
+
+#[cxx::bridge]
+mod ffi {
+    #[namespace = "seastar_ffi"]
+    unsafe extern "C++" {
+        type VoidFuture = crate::cxx_async_futures::VoidFuture;
+    }
+
+    #[namespace = "seastar_ffi::submit_to"]
+    unsafe extern "C++" {
+        include!("seastar/src/submit_to.hh");
+
+        unsafe fn submit_to(
+            _shard_id: u32,
+            closure: *mut u8,
+            caller: unsafe fn(*mut u8) -> VoidFuture,
+        ) -> VoidFuture;
+    }
+}
+
+/// Convert the pointer to Box<Func> and then call it, consuming it in the process.
+fn fn_once_caller<Func>(raw_func: *mut u8) -> VoidFuture
+where
+    Func: FnOnce() -> VoidFuture,
+{
+    unsafe { Box::from_raw(raw_func as *mut Func)() }
+}
+
+/// A helper function. We need to be able to name the type of a named closure.
+/// There is no `typeof` in Rust, so this is achieved by using a dummy paramter.
+const fn get_fn_once_caller<Func>(_: &Func) -> fn(*mut u8) -> VoidFuture
+where
+    Func: FnOnce() -> VoidFuture,
+{
+    fn_once_caller::<Func>
+}
+
+fn fn_once_dropper<Func>(raw_func: *mut u8)
+where
+    Func: FnOnce() -> VoidFuture,
+{
+    unsafe {
+        let _ = Box::from_raw(raw_func as *mut Func);
+    }
+}
+
+const fn get_fn_once_dropper<Func>(_: &Func) -> fn(*mut u8)
+where
+    Func: FnOnce() -> VoidFuture,
+{
+    fn_once_dropper::<Func>
+}
+
+/// Runs a function `func` on a `shard_id` shard.
+///
+/// # Example
+///
+/// ```rust
+/// #[seastar::test]
+/// async fn submit_to_example() {
+///     let ret = submit_to(0, || future::ready(42)).await;
+///     assert!(ret.is_ok());
+///     assert_eq!(ret.unwrap(), 42);
+/// }
+pub async fn submit_to<Func, Fut, Ret>(shard_id: u32, func: Func) -> cxx_async::CxxAsyncResult<Ret>
+where
+    Func: FnOnce() -> Fut + Send + 'static,
+    Fut: Future<Output = Ret> + 'static,
+    Ret: Send + std::fmt::Debug + 'static,
```

Comment:
Why the requirement for `std::fmt::Debug`?

Reply:
It is required by `std::result::Result::<T, E>::unwrap`.